envi/archs/aarch64/emu.py

```diff
@@ -0,0 +1,674 @@
+"""
+The initial aarch64 module.
+"""
+
+import sys
+import struct
+import logging
+
+import envi
+import envi.bits as e_bits
+from envi.const import *
+from envi.archs.aarch64.regs import *
+from envi.archs.aarch64 import A64Module
+
+logger = logging.getLogger(__name__)
+
+# CPU state (memory, regs inc SPSRs and banked registers)
+# CPU mode  (User, FIQ, IRQ, supervisor, Abort, Undefined, System)
+# 
+# instruction code
+# exception handler code
+# FIXME: SPSR handling is not certain.  
+
+# calling conventions
+
+## NO FREAKING CLUE YET....
+class A64Call(envi.CallingConvention):
+    arg_def = [(CC_REG, REG_X0), (CC_REG, REG_X1), (CC_REG, REG_X2), (CC_REG, REG_X3), 
+            (CC_REG, REG_X4), (CC_REG, REG_X5), (CC_REG, REG_X6), (CC_REG, REG_X7), 
+            (CC_STACK_INF, 8),]
+    retaddr_def = (CC_REG, REG_X30)
+    retval_def = (CC_REG, REG_X0)
+    flags = CC_CALLEE_CLEANUP
+    align = 8
+    pad = 0
+
+a64call = A64Call()
+###
+
+
+class CoProcEmulator:       # useful for prototyping, but should be subclassed
+    def __init__(self):
+        pass
+
+    def stc(self, parms):
+        logger.info("CoProcEmu: stc(%r)", parms)
+    def ldc(self, parms):
+        logger.info("CoProcEmu: ldc(%r)", parms)
+    def cdp(self, parms):
+        logger.info("CoProcEmu: cdp(%r)", parms)
+    def mcr(self, parms):
+        logger.info("CoProcEmu: mcr(%r)", parms)
+    def mcrr(self, parms):
+        logger.info("CoProcEmu: mcrr(%r)", parms)
+    def mrc(self, parms):
+        logger.info("CoProcEmu: mrc(%r)", parms)
+    def mrrc(self, parms):
+        logger.info("CoProcEmu: mrrc(%r)", parms)
+
+
+def _getRegIdx(idx, mode):
+    '''
+    Currently cheating and simply supporting only one mode and associated registers.
+    '''
+    if idx >= MAX_REGS:
+        return idx
+
+    ridx = idx
+    #ridx = idx + (mode*17)  # account for different banks of registers
+    #ridx = reg_table[ridx]  # magic pointers allowing overlapping banks of registers
+    return ridx
+
+def c0000(flags):
```

Comment:
Uhhh, could you give a comment or something describing these functions? Make it easier for me to figure out what's going on.